pkg/kinds/convert_test.go

```diff
@@ -0,0 +1,540 @@
+					},
+					Ports: []corev1.ServicePort{
+						{
+							Protocol:   corev1.ProtocolTCP,
+							Port:       80,
+							TargetPort: intstr.FromInt(9376),
+						},
+					},
+				},
+			},
+			scheme: coreScheme,
+			expected: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": Service().GroupVersion().String(),
+					"kind":       Service().Kind,
+					"metadata": map[string]interface{}{
+						"name":              "test-name",
+						"creationTimestamp": nil, // Added field
+					},
+					"spec": map[string]interface{}{
+						"selector": map[string]interface{}{
+							"app.kubernetes.io/name": "MyApp",
+						},
+						"ports": []interface{}{
+							map[string]interface{}{
+								"protocol":   "TCP",
+								"port":       int64(80),   // Type change
+								"targetPort": int64(9376), // Type change
+							},
+						},
+					},
+					"status": map[string]interface{}{ // Added field
+						"loadBalancer": map[string]interface{}{}, // Added field
+					},
+				},
+			},
+		},
+		{
+			name: "typed unpopulated GVK not in scheme",
+			object: &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-name",
+				},
+				Spec: corev1.ServiceSpec{
+					Selector: map[string]string{
+						"app.kubernetes.io/name": "MyApp",
+					},
+					Ports: []corev1.ServicePort{
+						{
+							Protocol:   corev1.ProtocolTCP,
+							Port:       80,
+							TargetPort: intstr.FromInt(9376),
+						},
+					},
+				},
+			},
+			scheme: emptyScheme,
+			expectedError: testutil.EqualError(
+				errors.Wrap(
+					runtime.NewNotRegisteredErrForType(emptyScheme.Name(),
+						reflect.TypeOf(corev1.Service{})),
+					"failed to lookup object type")),
+		},
+		{
+			name: "typed unpopulated GVK in scheme",
+			object: &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-name",
+				},
+				Spec: corev1.ServiceSpec{
+					Selector: map[string]string{
+						"app.kubernetes.io/name": "MyApp",
+					},
+					Ports: []corev1.ServicePort{
+						{
+							Protocol:   corev1.ProtocolTCP,
+							Port:       80,
+							TargetPort: intstr.FromInt(9376),
+						},
+					},
+				},
+			},
+			scheme: coreScheme,
+			expected: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": Service().GroupVersion().String(),
+					"kind":       Service().Kind,
+					"metadata": map[string]interface{}{
+						"name":              "test-name",
+						"creationTimestamp": nil, // Added field
+					},
+					"spec": map[string]interface{}{
+						"selector": map[string]interface{}{
+							"app.kubernetes.io/name": "MyApp",
+						},
+						"ports": []interface{}{
+							map[string]interface{}{
+								"protocol":   "TCP",
+								"port":       int64(80),   // Type change
+								"targetPort": int64(9376), // Type change
+							},
+						},
+					},
+					"status": map[string]interface{}{ // Added field
+						"loadBalancer": map[string]interface{}{}, // Added field
```

Comment:
Extracted this code to https://github.com/GoogleContainerTools/kpt-config-sync/pull/81

But I assume it's because they're marked as required in the scheme.

Reply:
Hmm. Service .status and .status.loadBalancer are both marked as optional/omitempty in the v1 scheme.

Reply:
I talked to the apimachinery team. It turns out this is just an artifact of how the unstructured conversion code works. You could consider it a bug but it's how it's been implemented. It basically ignores omitempty/optional on structs when recursing.

Reply:
It turns out this is how the Golanf json.Marshal work. So the reflection impl was made to match that for consistency. Even if I could fix the reflection impl, the Go son change would take months and may not even get approved due to reverse incompatibility.